### PR TITLE
Secure memory refactor

### PR DESCRIPTION
**Pull request by jgowdy**
This is the long awaited refactor of the poor abstractions of SecureMemory

### CODE REVIEW

File: csharp/SecureMemory/PlatformNative/SystemInterface.cs [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                    {
+                        systemInterface = new MacOSSystemInterfaceImpl();
+                    }
+                    else if (RuntimeInformation.IsOSPlatform(OSPlatform.Windows))
+                    {
+                        systemInterface = new WindowsSystemInterfaceImpl();
+                    }
+                    else
+                    {
+                        throw new PlatformNotSupportedException("Unknown platform");
```
Comment (rleverence-gd) [inline csharp/SecureMemory/PlatformNative/SystemInterface.cs:62]: Put the platform in the exception message? It's available via `RuntimeInformation.OSDescription`

File: csharp/SecureMemory/SecureMemory/ProtectedMemoryImpl/OpenSSL/OpenSSL11ProtectedMemoryAllocatorLP64.cs [hunk 2]
```diff
@@ -0,0 +1,193 @@
...
+            if (disposedValue)
+            {
+                throw new Exception("Called SetReadAccess on disposed LinuxOpenSSL11ProtectedMemoryAllocatorLP64");
+            }
+
+            Check.IntPtr(pointer, "SetReadAccess");
+
+            // Per page-protections aren't possible with the OpenSSL secure heap implementation
+            // Round up allocation size to nearest block size
+            // Debug.WriteLine($"SetReadAccess: Rounding length {length} to nearest blocksize");
```
Comment (rleverence-gd) [inline csharp/SecureMemory/SecureMemory/ProtectedMemoryImpl/OpenSSL/OpenSSL11ProtectedMemoryAllocatorLP64.cs:109]: Commented out code

File: csharp/SecureMemory/SecureMemory/ProtectedMemoryImpl/OpenSSL/OpenSSL11ProtectedMemoryAllocatorLP64.cs [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+        public void Dispose()
+        {
+            disposedValue = true;
+            ReleaseUnmanagedResources();
+            GC.SuppressFinalize(this);
+        }
+
+        private void ReleaseUnmanagedResources()
+        {
+            // TODO release unmanaged resources here
```
Comment (rleverence-gd) [inline csharp/SecureMemory/SecureMemory/ProtectedMemoryImpl/OpenSSL/OpenSSL11ProtectedMemoryAllocatorLP64.cs:190]: Is this a critical thing to have done in this PR?

File: csharp/SecureMemory/SecureMemory/ProtectedMemoryImpl/Windows/WindowsProtectedMemoryAllocatorLLP64.cs [hunk 4]
```diff
@@ -1,93 +1,114 @@
...
-        public void SetNoAccess(IntPtr pointer, ulong length)
+        public WindowsProtectedMemoryAllocatorLLP64(IConfiguration configuration, SystemInterface systemInterface, IMemoryEncryption memoryEncryption)
         {
-            length = AdjustLength(length);
+            this.systemInterface = systemInterface ?? throw new ArgumentNullException(nameof(systemInterface));
+            this.memoryEncryption = memoryEncryption;
+            encryptedMemoryBlockSize = memoryEncryption.GetEncryptedMemoryBlockSize();
 
-            if (!WindowsInterop.CryptProtectMemory(pointer, (UIntPtr)length, CryptProtectMemoryOptions.SAME_PROCESS))
+            /*
```
Comment (rleverence-gd) [inline csharp/SecureMemory/SecureMemory/ProtectedMemoryImpl/Windows/WindowsProtectedMemoryAllocatorLLP64.cs:22]: Commented out code.

**Review by rleverence-gd — CHANGES_REQUESTED** [review]

**Comment by jgowdy** [comment]
I think there's an issue here where I'm treating the IV field as an actual IV in OpenSSLCryptProtectMemory rather than the fact that in AEAD it's repurposed as a nonce field.  I think this class suffers from the dreaded nonce reuse for the same key issue.

File: csharp/SecureMemory/PlatformNative/OpenSSL/OpenSSLCryptoWindows.cs [hunk 5]
```diff
@@ -0,0 +1,329 @@
...
+        }
+
+        [DllImport(LibraryName, EntryPoint = "EVP_CIPHER_block_size", SetLastError = true)]
+        private static extern int _EVP_CIPHER_block_size(IntPtr e);
+
+        public int EVP_CIPHER_block_size(IntPtr e)
+        {
+            int blockSize = _EVP_CIPHER_block_size(e);
+
+            // BUG: EVP_CIPHER_block_size returns 1
```
[the remark this answers is not in the document]
Comment (jgowdy) [reply csharp/SecureMemory/PlatformNative/OpenSSL/OpenSSLCryptoWindows.cs:295]: I've figured this out, and this actually isn't a bug.  With AEAD ciphers like GCM there is no padding which is why EVP_CIPHER_block_size returns 1 byte.  With AEAD, if you encrypt 2 bytes, you get 2 bytes, there's no rounding to the next block size like with CBC.

**Comment by jgowdy** [comment]
Some block modes (like CBC) require the input to be split into blocks and the final block to be padded to the block size using a padding algorithm (e.g. add a special padding character). Other block modes (like CTR, CFB, OFB, CCM, EAX and GCM) do not require padding at all, because they perform XOR between portions of the plaintext and the internal cipher's state at each step.

**Comment by jgowdy** [comment]
I'm changing direction on this

**Pull request closed without merging**